src/liballoc/raw_vec.rs

```diff
@@ -571,15 +571,15 @@ impl<T> RawVec<T, Global> {
     ///
     /// # Safety
     ///
-    /// `shrink_to_fit(len)` must be called immediately prior to calling this function. This
-    /// implies, that `len` must be smaller than or equal to `self.capacity()`.
+    /// `shrink_to_fit(len)` must be called immediately prior to calling this function.
+    /// In particular, this implies that `len` must be smaller than or equal to `self.capacity()`
+    /// (but that is just a *necessary*, not a *sufficient*, condition).
```

Comment:
```suggestion
    /// * `len` must be greater than or equal to the capacity requested in a previous call
    /// * `len` must be smaller than or equal to `self.capacity()`
    ///
    /// Note, that the requested capacity and `self.capacity()` may not be the same, as
    /// an allocator may overallocate and return a greater memory block than requested.
```

Reply:
The thing mentioned in the note was the reason why I added a `len` parameter to the method by the way. It's also the reason why we don't have to special case the internal capacity field anymore.

Reply:
Thanks, I'll use that and edit some more!
"may not be the same" sounds like "must not be the same", so that is a bit confusing. Also, I think "most recent call" is the more common expression.

Reply:
> The thing mentioned in the note was the reason why I added a len parameter to the method by the way. It's also the reason why we don't have to special case the internal capacity field anymore.

So if I see it correctly, the difference between `self.cap` and `self.capacity()` only arises in case of `mem::size_of::<T>() == 0`. And in that case the size of the `Box` is zero anyway so it does not matter?

Reply:
I think your current wording of your latest commit is fine.

The unsafetyness is basically because `len * mem::size_of::<T>()` is passed to `AllocRef::dealloc` when dropping the `Box`, this must be a valid call.

Reply:
Yes, but I don't understand why `self.cap` vs `self.capacity()` was important, and why it is not relevant any more?

Reply:
> I don't understand why self.cap vs self.capacity() was important, and why it is not relevant any more?

It was only important as calls to `shrink_to_fit` have set `self.cap` in any case, even for ZSTs. Calling `into_box` then returned a boxed slice with the `boxed.len() == self.cap`. As `len` is now passed explicitly to `into_box`, `self.cap` isn't used here anymore. 

> So if I see it correctly, the difference between self.cap and self.capacity() only arises in case of mem::size_of::<T>() == 0. And in that case the size of the Box is zero anyway so it does not matter?

Yes, that's correct. When using `RawVec<ZST>`, `AllocRef` is called with a `layout.size() == 0`, and `self.capacity()` always return `usize::MAX`. Theoretically *any* length can be passed to `into_box` here. However, the safety constraint is still a little bit more conservative. For example calling `RawVec::<ZST>::with_capacity(10)` forbids calling `into_box(5)`, but this would be well defined. Personally I'd keep the safety clause as mentioned above, because then ZSTs and none-ZSTs are exchangeable.